src/Microsoft.ML.TensorFlow/TensorflowTransform.cs

```diff
@@ -111,34 +102,55 @@ private static TensorFlowTransform Create(IHostEnvironment env, ModelLoadContext
+
             // *** Binary format ***
+            // byte: indicator for frozen models
             // stream: tensorFlow model.
             // int: number of input columns
             // for each input column
             //   int: id of int column name
             // int: number of output columns
             // for each output column
             //   int: id of output column name
-            byte[] modelBytes = null;
-            if (!ctx.TryLoadBinaryStream("TFModel", r => modelBytes = r.ReadByteArray()))
-                throw env.ExceptDecode();
-            var numInputs = ctx.Reader.ReadInt32();
-            env.CheckDecode(numInputs > 0);
-            string[] inputs = new string[numInputs];
-            for (int j = 0; j < inputs.Length; j++)
-                inputs[j] = ctx.LoadNonEmptyString();
+            GetModelInfo(env, ctx, out string[] inputs, out string[] outputs, out bool isFrozen);
+            if (isFrozen)
+            {
+                byte[] modelBytes = null;
+                if (!ctx.TryLoadBinaryStream("TFModel", r => modelBytes = r.ReadByteArray()))
+                    throw env.ExceptDecode();
 
-            bool isMultiOutput = ctx.Header.ModelVerReadable >= 0x00010002;
-            var numOutputs = 1;
-            if (isMultiOutput)
-                numOutputs = ctx.Reader.ReadInt32();
+                return new TensorFlowTransform(env, LoadTFSession(env, modelBytes), inputs, outputs, isFrozen);
+            }
 
-            env.CheckDecode(numOutputs > 0);
-            var outputs = new string[numOutputs];
-            for (int j = 0; j < outputs.Length; j++)
-                outputs[j] = ctx.LoadNonEmptyString();
+            var tempDirPath = Path.GetFullPath(Path.Combine(Path.GetTempPath(), RegistrationName + "_" + Guid.NewGuid()));
+            TensorFlowUtils.CreateTempDirectory(tempDirPath);
+
+            var load = ctx.TryLoadBinaryStream("TFSavedModel", br =>
+            {
+                int count = br.ReadInt32();
+                for (int n = 0; n < count; n++)
+                {
+                    string relativeFile = br.ReadString();
+                    long fileLength = br.ReadInt64();
+
+                    string fullFilePath = Path.Combine(tempDirPath, relativeFile);
+                    string fullFileDir = Path.GetDirectoryName(fullFilePath);
+
+                    if (fullFileDir != tempDirPath)
+                    {
+                        TensorFlowUtils.CreateTempDirectory(fullFileDir);
+                    }
 
-            return new TensorFlowTransform(env, modelBytes, inputs, outputs);
+                    using (var fs = new FileStream(fullFilePath, FileMode.Create, FileAccess.Write))
+                    {
+                        long actualRead = br.BaseStream.CopyRange(fs, fileLength);
+                        env.Assert(actualRead == fileLength);
+                    }
+                }
+            });
+
+            var session = GetSession(env, tempDirPath);
```

Comment:
>var session = GetSession(env, tempDirPath); [](start = 12, length = 43)

Why can't the temp directory be deleted right after the session is loaded? Is it used for anything else?

Reply:
Ok, I see. it is used to save the model too.

---
In reply to: [219211241](https://github.com/dotnet/machinelearning/pull/853#discussion_r219211241) [](ancestors = 219211241)

Reply:
This directory is *mutable* model state.  It must remain.

---
In reply to: [219211991](https://github.com/dotnet/machinelearning/pull/853#discussion_r219211991) [](ancestors = 219211991,219211241)